ext/standard/assert.c

```diff
@@ -151,9 +149,12 @@ PHP_FUNCTION(assert)
 		zval args[4];
 		zval retval;
 		uint32_t lineno = zend_get_executed_lineno();
-		const char *filename = zend_get_executed_filename();
+		zend_string *filename = zend_get_executed_filename_ex();
+		if (UNEXPECTED(!filename)) {
+			filename = ZSTR_KNOWN(ZEND_STR_UNKNOWN_CAPITALIZED);
+		}
 
-		ZVAL_STRING(&args[0], SAFE_STRING(filename));
+		ZVAL_STR(&args[0], filename);
```

Comment:
I think there's a bug here. At this point you put the filename into args[0], without taking a copy. But further down you call the destructor on args[0], so you decrement the refcount with 1 (or destroy it upon reaching 0).
So that means that every time you call assert() the filename gets its refcount decremented with 1.

So executing the following PHP script (with USE_ZEND_ALLOC=0, USE_TRACKED_ALLOC=1) crashes:
```php
<?php

assert_options(ASSERT_CALLBACK, function ($f) {});
try {
    assert(false);
} catch (Throwable) {}
try {
    assert(false);
} catch (Throwable) {}
try {
    assert(false);
} catch (Throwable) {}
try {
    assert(false);
} catch (Throwable) {}
try {
    assert(false);
} catch (Throwable) {}
try {
    assert(false);
} catch (Throwable) {}
try {
    assert(false);
} catch (Throwable) {}
```

Reply:
Indeed, I was running tests with opcache which hid this, but ASAN was already complaining even without this new test.